File: voice_agent/server/tenant_context.py
"""
Tenant context extraction for the Voice Agent Python server.
Resolves tenant_id from JWT tokens or API keys in request headers.
"""

import os
import json
import base64
from typing import Optional

DEFAULT_TENANT_ID = "00000000-0000-0000-0000-000000000001"
VOICE_AGENT_API_KEY = os.getenv("VOICE_AGENT_API_KEY", "")


def extract_tenant_from_request(headers: dict) -> dict:
    """
    Extract tenant_id and agent_email from request headers.

    Supports:
      - Legacy VOICE_AGENT_API_KEY (maps to default tenant)
      - JWT Bearer tokens with tenant_id claim
      - x-tenant-id header (explicit override for service-to-service)

    Returns:
        {"tenant_id": str, "agent_email": str | None, "error": str | None}
    """
    # Check x-tenant-id header (service-to-service override)
    explicit_tenant = headers.get("x-tenant-id")
    if explicit_tenant:
        return {"tenant_id": explicit_tenant, "agent_email": None, "error": None}

    auth = headers.get("Authorization", "")
    if not auth.startswith("Bearer "):
        return {"tenant_id": DEFAULT_TENANT_ID, "agent_email": None, "error": "Missing Bearer token"}

    token = auth[7:]

    # Legacy key check
    if VOICE_AGENT_API_KEY and token == VOICE_AGENT_API_KEY:
        return {"tenant_id": DEFAULT_TENANT_ID, "agent_email": None, "error": None}

    # Try JWT decode (no verification here — verified at API gateway level)
    payload = _decode_jwt_payload(token)
    if payload:
        return {
            "tenant_id": payload.get("tenant_id", DEFAULT_TENANT_ID),
            "agent_email": payload.get("sub"),
            "error": None,
        }

    # Unknown token format — use default tenant for backward compat
    return {"tenant_id": DEFAULT_TENANT_ID, "agent_email": None, "error": None}


def _decode_jwt_payload(token: str) -> Optional[dict]:
    """Decode JWT payload without verification (just extract claims)."""
    try:
        parts = token.split(".")
        if len(parts) != 3:
            return None
        # Add padding
        payload_b64 = parts[1] + "=" * (4 - len(parts[1]) % 4)
        payload_bytes = base64.urlsafe_b64decode(payload_b64)
        return json.loads(payload_bytes)
    except Exception:
        return None

File: src/lib/tenant-context.ts
/**
 * Multi-tenant context management.
 * Sets and retrieves tenant_id for Supabase RLS isolation.
 */

import { createAdminClient } from "@/lib/supabase/admin";

const DEFAULT_TENANT_ID = "00000000-0000-0000-0000-000000000001";

/**
 * Set tenant context on a Supabase connection so RLS policies filter by tenant.
 * Must be called before any tenant-scoped query.
 */
export async function setTenantContext(tenantId: string) {
  const supabase = createAdminClient();
  await supabase.rpc("set_tenant_context", { p_tenant_id: tenantId });
  return supabase;
}

/**
 * Get a Supabase client with tenant context already set.
 * Convenience wrapper for the common pattern.
 */
export async function getTenantClient(tenantId: string) {
  return setTenantContext(tenantId);
}

/**
 * Extract tenant_id from a JWT token's claims.
 * Returns the default tenant if no tenant_id claim exists (backward compat).
 */
export function extractTenantId(jwtPayload: Record<string, unknown>): string {
  const tenantId = jwtPayload.tenant_id as string | undefined;
  return tenantId ?? DEFAULT_TENANT_ID;
}

/**
 * Resolve tenant_id from a request's Authorization header.
 * Supports both:
 *   - Per-tenant API keys (looked up in tenant_api_keys table)
 *   - JWT tokens with tenant_id claim (OAuth2 flow)
 *
 * Falls back to default tenant for backward compatibility during migration.
 */
export async function resolveTenantFromRequest(
  request: Request
): Promise<{ tenantId: string; agentEmail: string | null; error: string | null }> {
  const auth = request.headers.get("Authorization");
  if (!auth?.startsWith("Bearer ")) {
    return { tenantId: DEFAULT_TENANT_ID, agentEmail: null, error: "Missing Authorization header" };
  }

  const token = auth.slice(7);

  // Phase 1: Check against legacy VOICE_AGENT_API_KEY for backward compat
  const legacyKey = process.env.VOICE_AGENT_API_KEY;
  if (legacyKey && token === legacyKey) {
    return { tenantId: DEFAULT_TENANT_ID, agentEmail: null, error: null };
  }

  // Phase 2: Try to decode as JWT (OAuth2 tokens)
  // TODO: Implement JWT validation in Step 5 (OAuth2 server)
  // For now, fall back to legacy key check
  return { tenantId: DEFAULT_TENANT_ID, agentEmail: null, error: null };
}

export { DEFAULT_TENANT_ID };
